src/or/hs_service.c

```diff
@@ -1053,6 +1056,222 @@ load_service_keys(hs_service_t *service)
+    return -1;
+
+  /* Iterater through each line to parse each client name and public key.
+   * Each line consists of the client name and the public key separated by
+   * a space. */
+  while (s < end) {
+    eol = memchr(s, '\n', end-s);
+    if (!eol) eol = (char *) end;
+    sp = memchr(s, ' ', eol-s);
+    if (!sp) goto end;
+
+    /* If the space is at the beginning of the line, it means there is
+     * client name, and, if the space is at the end of the line, it means
+     * there is no public key. In either case, the function should fail */
+    if (sp == s || sp+1 == eol) goto end;
+
+    hs_service_authorized_client_t *client =
+      tor_malloc_zero(sizeof(hs_service_authorized_client_t));
+
+    client->client_name = tor_malloc(sp-s+1);
+    memcpy(client->client_name, s, sp-s);
+    client->client_name[sp-s] = '\0';
+
+    /* The public key starts at sp+1, so we do not need to add one to
+     * accommodate null terminator. */
+    char *encoded_pubkey = tor_malloc(eol-sp);
+    memcpy(encoded_pubkey, sp+1, eol-sp-1);
+    encoded_pubkey[eol-sp-1] = '\0';
+
+    if (curve25519_public_from_base64(&client->client_pk,
+                                      encoded_pubkey) < 0) {
+      goto end;
+    }
+    strmap_set(parsed_clients, client->client_name, client);
+    num_clients++;
+
+    tor_free(encoded_pubkey);
+
+    /* Go to the next line */
+    s = eol+1;
+  }
+
+  ret = num_clients;
+ end:
+  return ret;
+}
+
+/* Load all the client public keys for the given service. If there is no
+ * public key for any client, generate a new one for them. Return 0 on
+ * success else -1 on failure. */
+static int
+load_client_keys(hs_service_t *service)
+{
+  int ret = -1;
+  char buf[CURVE25519_BASE64_PADDED_LEN+1];
+  char *fname = NULL;
+  char *client_keys_dir_path = NULL;
+  char *client_keys_str = NULL;
+  const hs_service_config_t *config;
+  struct stat file_stat;
+  strmap_t *parsed_clients = strmap_new();
+
+  tor_assert(service);
+
+  config = &service->config;
+  /* If the client authorization is disabled, we do not need to do anything */
+  if (config->client_auth_type == HS_CLIENT_AUTH_TYPE_NULL) {
+    return 0;
+  }
+
+  /* Before calling this function, we already call load_service_keys to make
+   * sure that the directory exists with the right permission. So, if we
+   * cannot create a client secret key directory, we consider it as a bug. */
+  client_keys_dir_path = hs_path_from_filename(config->directory_path,
+                                               dir_client_seckeys);
+  if (BUG(hs_check_service_private_dir(get_options()->User,
+                                       client_keys_dir_path,
+                                       config->dir_group_readable, 1) < 0)) {
+    goto end;
+  }
+
+  /* If the client authorization is enabled, we must have a list of clients */
+  tor_assert(config->clients);
+
+  /* Load public keys from the file. If the file does not exist,
+   * parsed_clients will be empty. */
```

Comment:
I wonder if having both `client_authorized_pubkeys` and `client_authorized_privkeys/alice.privkey` is confusing to people. It's not really clear which key should be sent to the client by the service, and I guess that many opeartors would think that  they need to send the pubkey since they are not used to sending privkeys around. (Case in point: I was confused when setting up the service/client for testing)

Suggestion:
Perhaps we could merge these two designs into one. Instead of having a separate `pubkeys` file and a separate `.privkey` file, we could only have ` client_authorized_keys/alice.key` and `client_authorized_keys/bob.key` and `client_authorized_keys/charlie.key`. That file could contain either Alice's pubkey (if the keypair was generated by alice and sent to the service), or it could contain both Alice's pubkey and private key (if the keypair was generated by the service). In any case, the service operator would just need to send the key file to Alice. And Tor, instead of checking the `pubkeys` file, it could check whether Alice has a .key` file.

What do you think? Do you think this could simplify things or not?

Reply:
Why do we need to have one file for one client?

Similar to `authorized_keys` file in SSH, if we have one file, we can see both client name and public key of all clients in the same file.

Reply:
Another option! To prevent the issue about copying private key to the client, I think maybe we can do the same thing as SSH. That is, we will not let the service generate a keypair by itself, but if it really wants to, it needs to use a command line to generate it for each client.

What do you think?

Reply:
Hmm, I think we originally came up with one file per client, so that it's easier and more intuitive for the service to send the right file to the right client (i.e. to avoid copy-pasta fails, searching a big file for a client name,. etc.).

The main difference I can think with SSH here is that in the SSH case both the client and the service are tech people who know how to edit configuration files, whereas in our case the client might just be a web user. And asking the web user to generate credentials for a website and send it to the website owner might be a bit too much. That's why I think that preserving the "service sends key to the client" interface is a good idea from a UX perspective.

Perhaps the time to get in touch with the UX team is getting closer :)